lib/pgauge_exchange.cu

```diff
@@ -3,358 +3,258 @@
 
-    #ifndef GPU_COMMS
-      comm_wait(mh_recv_fwd[d]);
-      cudaMemcpyAsync(recvg_d[d], recvg[d], bytes[d], cudaMemcpyHostToDevice, GFStream[1]);
-    #endif
+        cudaMemcpyAsync(send[d], send_d[d], bytes[d], cudaMemcpyDeviceToHost, GFStream[0]);
+        cudaMemcpyAsync(sendg[d], sendg_d[d], bytes[d], cudaMemcpyDeviceToHost, GFStream[1]);
 
-      #ifdef GPU_COMMS
-      comm_wait(mh_recv_fwd[d]);
-      #endif
-      Kernel_UnPack<NElems, Float, Gauge, false> <<< grid[d], block[d], 0, GFStream[1] >>>
-	(faceVolumeCB[d], dataexarg, reinterpret_cast<complex<Float>*>(recvg_d[d]), parity, d, dir, X[d] - data.R()[d]);
+        qudaStreamSynchronize(GFStream[0]);
+        comm_start(mh_send_fwd[d]);
 
-      comm_wait(mh_send_back[d]);
-      comm_wait(mh_send_fwd[d]);
-      qudaStreamSynchronize(GFStream[0]);
-      qudaStreamSynchronize(GFStream[1]);
-    }
-    checkCudaError();
-    qudaDeviceSynchronize();
-  #endif
+        qudaStreamSynchronize(GFStream[1]);
+        comm_start(mh_send_back[d]);
 
-  }
+        comm_wait(mh_recv_back[d]);
+        cudaMemcpyAsync(recv_d[d], recv[d], bytes[d], cudaMemcpyHostToDevice, GFStream[0]);
 
+        arg.array = reinterpret_cast<complex<Float>*>(recv_d[d]);
+        arg.borderid = data.R()[d] - 1;
+        qudaLaunchKernel(Kernel_UnPack<recon, Float, false, decltype(arg)>, tp, GFStream[0], arg);
 
-  template<typename Float>
-  void PGaugeExchange( cudaGaugeField& data, const int dir, const int parity) {
+        comm_wait(mh_recv_fwd[d]);
+        cudaMemcpyAsync(recvg_d[d], recvg[d], bytes[d], cudaMemcpyHostToDevice, GFStream[1]);
 
+        arg.array = reinterpret_cast<complex<Float>*>(recvg_d[d]);
+        arg.borderid = X[d] - data.R()[d];
+        qudaLaunchKernel(Kernel_UnPack<recon, Float, false, decltype(arg)>, tp, GFStream[1], arg);
 
-    // Switching to FloatNOrder for the gauge field in order to support RECONSTRUCT_12
-    // Need to fix this!!
-    if ( data.isNative() ) {
-      if ( data.Reconstruct() == QUDA_RECONSTRUCT_NO ) {
-	typedef typename gauge_mapper<Float,QUDA_RECONSTRUCT_NO>::type G;
-	PGaugeExchange<Float, 18>(G(data), data, dir, parity);
-      } else if ( data.Reconstruct() == QUDA_RECONSTRUCT_12 ) {
-	typedef typename gauge_mapper<Float,QUDA_RECONSTRUCT_12>::type G;
-        PGaugeExchange<Float, 12>(G(data), data, dir, parity);
-      } else if ( data.Reconstruct() == QUDA_RECONSTRUCT_8 ) {
-	typedef typename gauge_mapper<Float,QUDA_RECONSTRUCT_8>::type G;
-        PGaugeExchange<Float, 8>(G(data), data, dir, parity);
-      } else {
-        errorQuda("Reconstruction type %d of gauge field not supported", data.Reconstruct());
+        comm_wait(mh_send_back[d]);
+        comm_wait(mh_send_fwd[d]);
+        qudaStreamSynchronize(GFStream[0]);
+        qudaStreamSynchronize(GFStream[1]);
       }
-    } else {
-      errorQuda("Invalid Gauge Order\n");
+      checkCudaError();
```

Comment:
Leaving this for now.  I think, once all API calls are moved their quda prefix equivalents, then we'll have no need at all for `checkCudaError` in the main library since all CUDA API calls will be guarded behind the abstraction, and equivalentl for other target backends.